### ISSUE TITLE
Как отменить отправку чека в налоговую ?

### ISSUE THREAD

**Issue by Aleks-Z** [shtrih-m/javapos_shtrih#119]
Здравствуйте.

Вопрос 
Если возможность отменить отправку данных в налоговую при неправильной печати чека ?

Если такое невозможно
Можно ли настроить отправку данных в налоговую так, чтобы передача данных будет осуществляться после успешной печати чека (без Exception) ?

**Comment by nyxiscoo1** [comment]
Добрый день.

Оно примерно так и работает. Печатается чек, сохраняется в ФН и потом ККТ его печатает. Если во время печати возникла ошибка - касса будет в режиме продолжения печати и чек потом распечатается.

А ситуация когда чек напечатался, но в драйвере было получено исключение можно сказать штатная(согласно закону Мерфи), может быть ситуация, что кассе была отправлена команда закрытия чека, но телефон сел/связь порвалась, а касса чек напечатала, эту ситуацию нельзя решить на уровне драйвера, ПО должно корректно ее обрабатывать.

Я ее обычно решаю с помощью машины состояний у документа, перед отправкой команды закрытия чека запоминаю текущий номер документа в ФН +1(это ожидаемый номер документа), номер ФН и выставляю соответственно состояние чеку "Перед закрытием". Если чек нормально закрывается, то состояние меняется на "Напечатан". 
Если же во время закрытия получено исключение, то при появлении связи с принтером проверяю, если есть документ в состоянии "Перед закрытием", если есть нужно запросить из ФН документ с сохраненным номером, если он есть и его сумма и тип совпадают с документом который пытались напечатать, то это он и можно считать документ пробитым, если нету такого документа в ФН или документ там другой, то значит он не пробит и нужно принять соответствующие действия.

**Comment by Aleks-Z** [comment]
У меня возникли вопросы 
- Каким методом можно получить текущий номер документа в ФН ?
- Каким методы нужно использовать для получения текущих данные о последнем документе (состояние, номер, сумма и тд) ?

**Comment by nyxiscoo1** [comment]
fsReadStatus и fsFindDocument.

**Comment by Aleks-Z** [comment]
>нужно принять соответствующие действия.

Правильно понимаю, что можно отменить чек в ФН ? И если да то какой метод надо использовать ?

**Comment by nyxiscoo1** [comment]
Да, можно, если он еще не закрыт, но лучше для этого использовать метод отмены чека обычный, т.к. если отменить чек только в ФН. но не в кассе, то произойдет рассинхронизация их состояния(касса будет думать, что в ФН чек открыт, а в ФН его отменили) все зависит от ККТ как она себя поведет в этом случае.

```Java
// проверяем, есть ли открытый документ в ФН
fsStatus = printer.fsReadStatus();

(fsStatus.getDocType().getValue() != FSDocType.FS_DOCTYPE_NONE)
     printer.fsCancelDocument(); // если есть отменяем
```

Если же чек был закрыт, то его отменить нельзя, это одна из основных функций ФН - предотвратить манипуляцию с историей пробитых чеков.

**Comment by Aleks-Z** [comment]
> для этого использовать метод отмены чека обычный

fsCancelDocument() - это метод отмены чека в обход терминала напрямую ФН ?
Если да, то как выглядит метод отмены обычный ?

**Comment by Aleks-Z** [comment]
В [FSDocumentInfo](https://github.com/shtrih-m/javapos_shtrih/blob/master/Source/Core/src/com/shtrih/fiscalprinter/command/FSDocumentInfo.java) не могу найти ничего похожего на состояние закрытого чека. Как его можно получить ?

**Comment by nyxiscoo1** [comment]
> В FSDocumentInfo не могу найти ничего похожего на состояние закрытого чека. Как его можно получить ?

[Тут написано](https://github.com/shtrih-m/javapos_shtrih/issues/121#issuecomment-734164283).

> fsCancelDocument() - это метод отмены чека в обход терминала напрямую в ФН ?

Да.

> Если да, то как выглядит метод отмены обычный ?

Предполагаю, что `printRecVoid`, если использовать методы Jpos.
Если посмотреть как я это делаю в [примере](https://github.com/shtrih-m/javapos_shtrih/blob/7f15228328090ddbb1dc0c3ef3798336c4b9e3a6/Source/android/tinyJavaPosTester/src/com/shtrih/tinyjavapostester/FFD105Documents.java#L504), то `resetPrinter`.

**Comment by Aleks-Z** [comment]
>Тут написано.

Тоесть получаеться алгоритм получения состояния чека
- Найти документ через 

```
val cmd = FSFindDocument(printer.sysPassword, docNumber)
printer.fsFindDocument(cmd)
```
- Откастовать к FSDocumentReceipt через 
```
val doc = Unchecked.cast<FSDocumentReceipt>(cmd.document!!)`
```
- Получить тип чека через 
```
doc.type
```
- Сравнить с константами из SmFptrConst
```
    public static final int RECEIPT_STATE_UNKNOWN = 0;
    public static final int RECEIPT_STATE_OPENED = 1;
    public static final int RECEIPT_STATE_CLOSED = 2;
    public static final int RECEIPT_STATE_CANCELLED = 3;
```

**Comment by Aleks-Z** [comment]
```
doc.type 
```
По смыслу не очень вяжеться с сотояние чека

**Comment by nyxiscoo1** [comment]
Это не состояние, а тип документа: открытие смены, закрытие смены, чек, чек коррекции, отчет о состоянии расчетов. Его нужно проверить перед кастом.
Чек появляется в ФН после успешного закрытия, до этого его найти в ФН будет невозможно.
Как проверить, что в ФН открыт документ и отменить его при рассинхронизации состоянии кассы и ФН я уже писал выше.
